internal/agent/device/osimage.go

```diff
@@ -0,0 +1,90 @@
+package device
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/agent/device/status"
+	"github.com/flightctl/flightctl/internal/container"
+	"github.com/flightctl/flightctl/pkg/executer"
+	"github.com/sirupsen/logrus"
+	"k8s.io/klog/v2"
+)
+
+const (
+	RebootingReason       = "Rebooting"
+	OsImageDegradedReason = "OSImageControllerDegraded"
+)
+
+type OSImageController struct {
+	bootc         *container.BootcCmd
+	statusManager status.Manager
+	log           *logrus.Logger
+	logPrefix     string
+}
+
+func NewOSImageController(
+	executor executer.Executer,
+	statusManager status.Manager,
+	log *logrus.Logger,
+	logPrefix string,
+) *OSImageController {
+	return &OSImageController{
+		bootc:     container.NewBootcCmd(executor),
+		statusManager: statusManager,
+		log:       log,
+		logPrefix: logPrefix,
+	}
+}
+
+func (c *OSImageController) Sync(ctx context.Context, desired *v1alpha1.RenderedDeviceSpec) error {
+	klog.V(4).Infof("%s syncing device image", c.logPrefix)
+	defer klog.V(4).Infof("%s finished syncing device image", c.logPrefix)
+
+	err := c.ensureImage(ctx, desired)
+	if err != nil {
+		if err := c.statusManager.UpdateConditionError(ctx, OsImageDegradedReason, err); err != nil {
+			klog.Errorf("Failed to update condition: %v", err)
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (c *OSImageController) ensureImage(ctx context.Context, desired *v1alpha1.RenderedDeviceSpec) error {
+	if desired.Os == nil {
+		klog.V(4).Infof("%s device os image is nil", c.logPrefix)
+		return nil
+	}
+
+	host, err := c.bootc.Status(ctx)
+	if err != nil {
+		return err
+	}
+
+	if host.Status.Booted.Image.Image.Image == desired.Os.Image {
+		klog.V(4).Infof("Host is already running os image %s", desired.Os.Image)
+		return nil
+	}
+
+	image := desired.Os.Image
+	c.log.Infof("Switching to os image %s", image)
+	os := *desired.Os
+	if err := c.bootc.Switch(ctx, os.Image); err != nil {
```

Comment:
How long do Switch and Apply take approximately?  Will it look like the agent is down for a long time?

Reply:
swithc takes a bit because it pulls down the image but this happens during runtime. It could in theory take a very long time. Apply is just a reboot I didn't see a long duration on that but it has not been calculated. We should be able to track this with metrics.

Reply:
put it in a go routine and will log every 30s  for now

Reply:
This will block status updates so we can consider that in a followup as well

Reply:
OK great